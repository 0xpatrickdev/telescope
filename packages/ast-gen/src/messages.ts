import * as t from '@babel/types';
import { Mutation } from './types';

export const addMsgMethod = ({ methodName, typeUrl, TypeName, methodToCall }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    {
      type: 'Identifier',
      name: 'value',
      typeAnnotation: {
        type: 'TSTypeAnnotation',
        typeAnnotation: {
          type: 'TSTypeReference',
          // @ts-ignore
          typeName: {
            type: 'Identifier',
            name: TypeName
          }
        }
      }
    }
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([
        t.objectProperty(
          t.identifier('typeUrl'),
          t.stringLiteral(typeUrl)
        ),
        t.objectProperty(
          t.identifier('value'),
          t.callExpression(
            t.memberExpression(
              t.identifier(TypeName),
              t.identifier(methodToCall)
            ),
            [t.identifier('value')]
          )
        )
      ])
    )]));
};

export const addFromJSONMethod = ({ methodName, typeUrl, TypeName }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    {
      type: 'Identifier',
      name: 'value',
      typeAnnotation: {
        type: 'TSTypeAnnotation',
        typeAnnotation: {
          type: 'TSTypeReference',
          // @ts-ignore
          typeName: {
            type: 'Identifier',
            name: 'any'
          }
        }
      }
    }
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([
        t.objectProperty(
          t.identifier('typeUrl'),
          t.stringLiteral(typeUrl)
        ),
        t.objectProperty(
          t.identifier('value'),
          t.callExpression(
            t.memberExpression(
              t.identifier(TypeName),
              t.identifier('fromJSON')
            ),
            [t.identifier('value')]
          )
        )
      ])
    )]));
};

export const addFromPartialMethod = ({ methodName, typeUrl, TypeName }) => {
  return addMsgMethod({ methodName, typeUrl, TypeName, methodToCall: 'fromPartial' });
};

export const addToJSONMethod = ({ methodName, typeUrl, TypeName }) => {
  return addMsgMethod({ methodName, typeUrl, TypeName, methodToCall: 'toJSON' });
};


export const addJsonMethod = ({ methodName, typeUrl, TypeName }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    {
      type: 'Identifier',
      name: 'value',
      typeAnnotation: {
        type: 'TSTypeAnnotation',
        typeAnnotation: {
          type: 'TSTypeReference',
          // @ts-ignore
          typeName: {
            type: 'Identifier',
            name: TypeName
          }
        }
      }
    }
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([
        t.objectProperty(
          t.identifier('typeUrl'),
          t.stringLiteral(typeUrl)
        ),
        t.objectProperty(
          t.identifier('value'),
          t.identifier('value'),
          false,
          true
        )
      ])
    )]));
};

export const addEncodedMethod = ({ methodName, typeUrl, TypeName }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    {
      type: 'Identifier',
      name: 'value',
      typeAnnotation: {
        type: 'TSTypeAnnotation',
        typeAnnotation: {
          type: 'TSTypeReference',
          // @ts-ignore
          typeName: {
            type: 'Identifier',
            name: TypeName
          }
        }
      }
    }
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([

        t.objectProperty(
          t.identifier('type_url'),
          t.stringLiteral(typeUrl)
        ),

        t.objectProperty(
          t.identifier('value'),
          t.callExpression(
            t.memberExpression(
              t.callExpression(
                t.memberExpression(
                  t.identifier(TypeName),
                  t.identifier('encode')
                ),
                [
                  t.identifier('value')
                ]
              ),
              t.identifier('finish')
            ),
            []
          )
        )
      ])
    )]));
};


export const messages = (mutations: Mutation[]) => t.exportNamedDeclaration(t.variableDeclaration('const', [
  t.variableDeclarator(t.identifier('messages'), t.objectExpression(
    mutations.map(mutation => addFromPartialMethod(mutation))
  ))]));

export const encoded = (mutations: Mutation[]) => t.exportNamedDeclaration(t.variableDeclaration('const', [
  t.variableDeclarator(t.identifier('encoded'), t.objectExpression(
    mutations.map(mutation => addEncodedMethod(mutation))
  ))]));
