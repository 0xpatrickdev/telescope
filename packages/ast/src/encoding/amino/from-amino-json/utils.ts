import * as t from '@babel/types';
import { BILLION, memberExpressionOrIdentifier, memberExpressionOrIdentifierAminoCasing, shorthandProperty } from '../../../utils';
import { FromAminoParseField, fromAminoParseField } from './index'
import { protoFieldsToArray } from '../utils';
import { getOneOfs, getFieldOptionality } from '../../proto';

export const fromAmino = {
    defaultType(args: FromAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.field.name), memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn))
    },

    string(args: FromAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.field.name), memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn))
    },

    long(args: FromAminoParseField) {
        args.context.addUtil('Long');

        return t.objectProperty(t.identifier(args.field.name),
            t.callExpression(
                t.memberExpression(
                    t.identifier('Long'),
                    t.identifier('fromString')
                ),
                [
                    memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                ]
            ));
    },

    duration(args: FromAminoParseField) {
        args.context.addUtil('Long');

        const value = t.objectExpression(
            [
                t.objectProperty(t.identifier('seconds'), t.callExpression(
                    t.memberExpression(t.identifier('Long'), t.identifier('fromNumber')), [
                    t.callExpression(
                        t.memberExpression(
                            t.identifier('Math'),
                            t.identifier('floor')
                        ),
                        [
                            t.binaryExpression('/',
                                t.callExpression(
                                    t.identifier('parseInt'),
                                    [
                                        memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                                    ]
                                ),
                                BILLION
                            )
                        ]
                    )
                ]
                )),
                t.objectProperty(
                    t.identifier('nanos'),
                    t.binaryExpression('%',
                        t.callExpression(
                            t.identifier('parseInt'),
                            [
                                memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                            ]
                        ),
                        BILLION
                    )
                )
            ]
        );
        return t.objectProperty(t.identifier(args.field.name), value);
    },

    height(args: FromAminoParseField) {
        args.context.addUtil('Long');

        return t.objectProperty(
            t.identifier(args.field.name),
            t.conditionalExpression(
                memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn),
                t.objectExpression([
                    t.objectProperty(t.identifier('revisionHeight'),
                        t.callExpression(
                            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
                            [
                                t.logicalExpression(
                                    '||',
                                    t.memberExpression(
                                        memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn),
                                        t.identifier(args.context.options.aminoCasingFn('revision_height'))
                                    ),
                                    t.stringLiteral('0')
                                ),
                                t.booleanLiteral(true)
                            ])
                    ),
                    //
                    t.objectProperty(t.identifier('revisionNumber'),
                        t.callExpression(
                            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
                            [
                                t.logicalExpression(
                                    '||',
                                    t.memberExpression(
                                        memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn),
                                        t.identifier(args.context.options.aminoCasingFn('revision_number'))
                                    ),
                                    t.stringLiteral('0')
                                ),
                                t.booleanLiteral(true)
                            ])
                    )
                ]),
                t.identifier('undefined')
            )
        )
    },

    enum({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const enumFunction = context.lookupEnumFromJson(field, currentProtoPath);
        const value = t.callExpression(
            t.identifier(enumFunction), [
            memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn)
        ]);
        return t.objectProperty(t.identifier(field.name), value);
    },

    enumArray({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const enumFunction = context.lookupEnumFromJson(field, currentProtoPath);
        const value = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier('el')
                    ],
                    t.callExpression(
                        t.identifier(enumFunction),
                        [
                            t.identifier('el')
                        ]
                    )
                )
            ]
        );
        return t.objectProperty(t.identifier(field.name), value);
    },

    type({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const parentField = field;
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;
            return fromAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [...scope],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality?
            })
        });
        return t.objectProperty(t.identifier(field.name),
            t.objectExpression(
                properties
            )
        );
    },

    arrayFrom(args: FromAminoParseField) {
        return t.objectProperty(t.identifier(args.field.name),
            t.callExpression(
                t.memberExpression(
                    t.identifier('Array'),
                    t.identifier('from')
                ),
                [
                    memberExpressionOrIdentifierAminoCasing(args.scope, args.context.options.aminoCasingFn)
                ]
            ));
    },

    typeArray({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField) {
        const variable = 'el' + nested;
        const parentField = field;
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;

            return fromAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [variable],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality?
            })
        });

        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    t.objectExpression(
                        properties
                    )
                )
            ]
        );

        return t.objectProperty(t.identifier(field.name),
            expr
        );
    },


    scalarArray({ context, field, currentProtoPath, scope, nested, isOptional }: FromAminoParseField, arrayTypeAstFunc: Function) {
        const variable = 'el' + nested;

        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifierAminoCasing(scope, context.options.aminoCasingFn),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    arrayTypeAstFunc(variable)
                )
            ]
        );

        return t.objectProperty(t.identifier(context.options.aminoCasingFn(field.name)),
            expr
        );
    },

    pubkey(args: FromAminoParseField) {
        args.context.addUtil('toBase64');
        args.context.addUtil('encodeBech32Pubkey');

        return t.objectProperty(
            t.identifier(args.field.name),
            t.callExpression(
                t.identifier('encodeBech32Pubkey'),
                [
                    t.objectExpression([
                        t.objectProperty(
                            t.identifier('type'),
                            t.stringLiteral('tendermint/PubKeySecp256k1')
                        ),
                        t.objectProperty(
                            t.identifier('value'),
                            t.callExpression(
                                t.identifier('toBase64'),
                                [
                                    t.memberExpression(
                                        t.identifier('pubkey'),
                                        t.identifier('value')
                                    )
                                ]
                            )
                        )
                    ]),
                    // TODO how to manage this?
                    // 1. options.prefix
                    // 2. look into prefix and how it's used across chains
                    // 3. maybe AminoConverter is a class and has this.prefix!
                    t.stringLiteral('cosmos')
                ]
            )
        )
    }
};


export const arrayTypes = {
    long(varname: string) {
        return t.callExpression(
            t.memberExpression(t.identifier('Long'), t.identifier('fromString')),
            [
                t.identifier(varname)
            ]
        )
    }
}
