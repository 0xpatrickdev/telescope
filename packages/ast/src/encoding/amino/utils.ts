import * as t from '@babel/types';
import { ProtoType } from '@cosmology/types';

export { getTypeUrlWithPkgAndName, getTypeUrl, getAminoTypeName } from '@cosmology/utils'

export const arrayTypeNDim = (body, n) => {
    if (!n || n <= 1) return t.tsArrayType(body);
    return t.tsArrayType(
        arrayTypeNDim(body, n - 1)
    );
};

export const protoFieldsToArray = (proto: ProtoType) => {
    return Object.keys(proto.fields).map(name => {
        return {
            name,
            ...proto.fields[name]
        };
    })
}