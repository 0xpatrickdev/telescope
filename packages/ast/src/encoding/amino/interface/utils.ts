import * as t from '@babel/types';
import { memberExpressionOrIdentifierAminoCasing, shorthandProperty } from '../../../utils';
import { AminoOptions } from '../types';
import { arrayTypeNDim, BILLION } from '../utils';
import { getTypeFromContext, protoFieldsToArray } from '../utils';
import { getTSTypeFromProtoType, ProtoField } from '../../types';
import { RenderAminoField, renderAminoField } from '.';

export const aminoInterface = {
    defaultType(field: ProtoField, options: AminoOptions, ast: any) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(ast)
        );
    },
    long(field: ProtoField, options: AminoOptions) {
        // longs become strings...
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(t.tSStringKeyword())
        )
    },
    height(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier('AminoHeight'))
            )
        );
    },
    enum(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(t.tSNumberKeyword())
        );
    },
    enumArray(field: ProtoField, options: AminoOptions) {
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(arrayTypeNDim(t.tSNumberKeyword(), 1))
        );
    },
    type({ context, field, options }: RenderAminoField) {
        const Type = getTypeFromContext(field, context);
        const properties = protoFieldsToArray(Type).map(field => {
            return renderAminoField({
                context,
                field,
                options
            })
        });

        // 
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                t.tsTypeLiteral(
                    properties
                )
            )
        );
    },
    typeArray({ context, field, options }: RenderAminoField) {
        const Type = getTypeFromContext(field, context);
        const properties = protoFieldsToArray(Type).map(field => {
            return renderAminoField({
                context,
                field,
                options
            })
        });

        // 
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                arrayTypeNDim(t.tsTypeLiteral(
                    properties
                ), 1)
            )
        );
    },

    array(field: ProtoField, options: AminoOptions) {
        // TODO write test case 

        // return t.tsPropertySignature(
        //     t.identifier(options.aminoCasingFn(field.name)),
        //     t.tsTypeAnnotation(
        //         arrayTypeNDim(t.tsTypeLiteral(
        //             properties
        //         ), 1)
        //     )
        // );
        return t.tsPropertySignature(
            t.identifier(options.aminoCasingFn(field.name)),
            t.tsTypeAnnotation(
                arrayTypeNDim(
                    getTSTypeFromProtoType(field.type),
                    1
                )
            )
        );
    }
}