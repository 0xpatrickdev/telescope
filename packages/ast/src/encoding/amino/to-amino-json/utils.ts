import * as t from '@babel/types';
import { BILLION, memberExpressionOrIdentifier, shorthandProperty } from "../../../utils";
import { protoFieldsToArray } from '../utils';
import { ToAminoParseField, toAminoParseField } from './index'
import { getOneOfs, getFieldOptionality } from '../../proto';

export const toAmino = {
    defaultType(args: ToAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), memberExpressionOrIdentifier(args.scope))
    },

    long(args: ToAminoParseField) {
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)),
            t.callExpression(
                t.memberExpression(memberExpressionOrIdentifier(args.scope), t.identifier('toString')),
                [])
        )
    },

    string(args: ToAminoParseField) {
        if (args.field.name === args.context.options.aminoCasingFn(args.field.name) && args.scope.length === 1) {
            return shorthandProperty(args.field.name);
        }
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), memberExpressionOrIdentifier(args.scope))
    },

    duration(args: ToAminoParseField) {
        const exp = t.binaryExpression(
            '*',
            memberExpressionOrIdentifier(args.scope),
            BILLION
        );
        exp.extra = { parenthesized: true };
        const value = t.callExpression(
            t.memberExpression(
                exp,
                t.identifier('toString')
            ),
            []
        )
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), value);
    },

    height(args: ToAminoParseField) {
        args.context.addUtil('omitDefault');

        const value = t.objectExpression(
            [
                t.objectProperty(
                    t.identifier(args.context.options.aminoCasingFn('revision_height')),
                    t.optionalCallExpression(
                        t.optionalMemberExpression(
                            t.callExpression(
                                t.identifier('omitDefault'),
                                [
                                    t.memberExpression(
                                        memberExpressionOrIdentifier(args.scope),
                                        t.identifier('revisionHeight')
                                    )
                                ]
                            ),
                            t.identifier('toString'),
                            false,
                            true
                        ),
                        [],
                        false
                    )
                ),
                //
                t.objectProperty(
                    t.identifier(args.context.options.aminoCasingFn('revision_number')),
                    t.optionalCallExpression(
                        t.optionalMemberExpression(
                            t.callExpression(
                                t.identifier('omitDefault'),
                                [
                                    t.memberExpression(
                                        memberExpressionOrIdentifier(args.scope),
                                        t.identifier('revisionNumber')
                                    )
                                ]
                            ),
                            t.identifier('toString'),
                            false,
                            true
                        ),
                        [],
                        false
                    )
                )
            ]
        );

        const cond = t.conditionalExpression(
            memberExpressionOrIdentifier(args.scope),
            value,
            t.objectExpression([])
        );

        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), cond);
    },

    coin(args: ToAminoParseField) {
        const value = t.objectExpression([
            t.objectProperty(t.identifier('denom'), t.memberExpression(
                memberExpressionOrIdentifier(args.scope),
                t.identifier('denom'),
            )),
            t.objectProperty(
                t.identifier('amount'),
                t.callExpression(
                    t.memberExpression(
                        t.callExpression(
                            t.memberExpression(
                                t.identifier('Long'),
                                t.identifier('fromNumber')
                            ),
                            [
                                t.memberExpression(
                                    memberExpressionOrIdentifier(args.scope),
                                    t.identifier('amount')
                                )
                            ]
                        ),
                        t.identifier('toString')
                    ),
                    []
                )
            )
        ]);
        return t.objectProperty(t.identifier(args.context.options.aminoCasingFn(args.field.name)), value);
    },

    type({ context, field, currentProtoPath, scope, nested, isOptional }: ToAminoParseField) {
        /// TODO (can this be refactored out? e.g. no recursive calls in this file?)
        /// BEGIN
        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const parentField = field;
        const oneOfs = getOneOfs(Type);
        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;
            return toAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [...scope],
                nested,
                isOptional // TODO how to handle nested optionality
            })
        });
        /// END 
        return t.objectProperty(t.identifier(context.options.aminoCasingFn(field.name)),
            t.objectExpression(
                properties
            )
        );
    },


    typeArray({ context, field, currentProtoPath, scope, nested, isOptional }: ToAminoParseField) {
        const variable = 'el' + nested;

        if (field.parsedType.type !== 'Type') {
            throw new Error('Arrays only support types[Type] right now.');
        }

        const Type = context.getTypeFromCurrentPath(field, currentProtoPath);
        const parentField = field;
        const oneOfs = getOneOfs(Type);

        const properties = protoFieldsToArray(Type).map(field => {
            const isOneOf = oneOfs.includes(field.name);
            const isOptional = getFieldOptionality(field, isOneOf);

            if (parentField.import) currentProtoPath = parentField.import;

            return toAminoParseField({
                context,
                field,
                currentProtoPath,
                scope: [variable],
                nested: nested + 1,
                isOptional // TODO how to handle nested optionality
            });
        });


        const expr = t.callExpression(
            t.memberExpression(
                memberExpressionOrIdentifier(scope),
                t.identifier('map')
            ),
            [
                t.arrowFunctionExpression(
                    [
                        t.identifier(variable)
                    ],
                    t.objectExpression(
                        properties
                    )
                )
            ]
        );

        return t.objectProperty(t.identifier(context.options.aminoCasingFn(field.name)),
            expr
        );
    },

    pubkey(args: ToAminoParseField) {
        args.context.addUtil('fromBase64');
        args.context.addUtil('decodeBech32Pubkey');

        return t.objectProperty(
            t.identifier(args.field.name),
            t.objectExpression([
                t.objectProperty(
                    t.identifier('typeUrl'),
                    t.stringLiteral('/cosmos.crypto.secp256k1.PubKey')
                ),
                t.objectProperty(
                    t.identifier('value'),
                    t.callExpression(
                        t.identifier('fromBase64'),
                        [
                            t.memberExpression(
                                t.callExpression(
                                    t.identifier('decodeBech32Pubkey'),
                                    [
                                        t.identifier(args.field.name)
                                    ]
                                ),
                                t.identifier('value')
                            )

                        ]
                    )
                )
            ])
        )

    }


};
