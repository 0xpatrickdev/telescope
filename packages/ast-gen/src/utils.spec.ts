import { encoded } from './messages';
import { recursiveNamespace } from './utils';
import generate from '@babel/generator';
import { mutations } from './__fixtures__';

const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}

it('recursiveNamespace', async () => {
    expectCode(
        recursiveNamespace(['osmosis', 'gamm', 'v1beta', 'pools'].reverse(), [encoded(mutations)])[0]
    );
});
