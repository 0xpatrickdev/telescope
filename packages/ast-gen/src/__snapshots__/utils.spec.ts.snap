// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recursiveNamespace 1`] = `
"export namespace osmosis {
  export namespace gamm {
    export namespace v1beta {
      export namespace pools {
        export const encoded = {
          joinPool(value: MsgJoinPool) {
            return {
              type_url: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
              value: MsgJoinPool.encode(value).finish()
            };
          },

          exitPool(value: MsgExitPool) {
            return {
              type_url: \\"/cosmos.pools.transfer.v1.MsgExitPool\\",
              value: MsgExitPool.encode(value).finish()
            };
          }

        };
      }
    }
  }
}"
`;
